File: ds_5_tree/tree_2_traversal/L103-m.cpp


//leetcode 103 二叉树锯齿形遍历

#include <iostream>
#include <cassert>
#include <queue>
#include <vector>
using namespace std ;

struct TreeNode{
    int val ;
    TreeNode* left ;
    TreeNode* right ;
    TreeNode(int x): val(x),left(NULL),right(NULL){}
};

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(root ==  NULL) return res;
        queue<TreeNode*> q;
        q.push(root);
        int level = 0;
        while(!q.empty()){
            vector<int> tmp;
            int sz = q.size();
            ++level;
            for(int i=0; i<sz; ++i){
                TreeNode* cur = q.front();
                q.pop();
                tmp.push_back(cur->val);
                if(cur->left)
                    q.push(cur->left);
                if(cur->right)
                    q.push(cur->right);
            }
            if (level%2 == 0)
                reverse(tmp.begin(), tmp.end());
            res.push_back(tmp);
        }
        return res;
    }
};

File: python-algorithm/ds_5_tree/tree_2_traversal/L103-m.py
# -*- coding: utf-8 -*-


class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []
        if not root:
            return res
        que = []
        level = 0
        que.append(root)
        while que!= []:
            level += 1
            tmp = []
            for _ in range(len(que)):
                cur = que.pop(0)
                tmp.append(cur.val)
                if cur.left:
                    que.append(cur.left)
                if cur.right:
                    que.append(cur.right)
            if level%2 == 0:
                res.append(tmp[::-1])
            else:
                res.append(tmp)
        return res